### PR TITLE
[tune] logger refactor part 2: Add SyncerCallback

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

Second split of PR ##11699, builds on #11746

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tune/utils/callback.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+def create_default_callbacks(callbacks: Optional[List[Callback]],
+                             sync_config: SyncConfig):
+
+    callbacks = callbacks or []
+    has_syncer_callback = False
+
+    # Check if we have a CSV and JSON logger
+    for i, callback in enumerate(callbacks):
+        if isinstance(callback, SyncerCallback):
+            has_syncer_callback = True
```
Comment (richardliaw) [inline python/ray/tune/utils/callback.py:18]: comment needs to be updated?

also, consider:
```
has_syncer_callback = any(isinstance(c, SyncerCallback) for c in callbacks)

**Comment by richardliaw** [comment]
@krfricke can you quickly fix conflicts?

File: python/ray/tune/syncer.py [hunk 2]
```diff
@@ -355,3 +363,88 @@ def get_node_syncer(local_dir, remote_dir=None, sync_function=None):
...
+                    raise e
+            if not issubclass(trial.get_trainable_cls(), DurableTrainable):
+                if not os.path.exists(checkpoint.value):
+                    raise TuneError("Trial {}: Checkpoint path {} not "
+                                    "found after successful sync down.".format(
+                                        trial, checkpoint.value))
+
+    def on_trial_start(self, iteration: int, trials: List["Trial"],
+                       trial: "Trial", **info):
+        self._create_trial_syncer(trial)
```
Comment (richardliaw) [inline python/ray/tune/syncer.py:430]: should this be `self._get_trial_syncer`?

**Review by richardliaw — APPROVED** [review]
this is great!

File: python/ray/tune/syncer.py (hunk 2 above)
Comment (krfricke) [reply python/ray/tune/syncer.py:430]: Yes! Thanks for spotting that!

File: src/ray/raylet/raylet.cc [hunk 3]
```diff
@@ -69,17 +69,17 @@ Raylet::Raylet(boost::asio::io_service &main_service, const std::string &socket_
...
+          main_service, self_node_id_, object_manager_config, object_directory_,
+          [this](const ObjectID &object_id, const std::string &spilled_url,
+                 std::function<void(const ray::Status &)> callback) {
+            node_manager_.GetLocalObjectManager().AsyncRestoreSpilledObject(
+                object_id, spilled_url, callback);
+          },
+          [this](int64_t num_bytes_required) {
+            return node_manager_.GetLocalObjectManager().SpillObjectsOfSize(
+                num_bytes_required);
+          }),
```
Comment (krfricke) [inline src/ray/raylet/raylet.cc:82]: This got auto linted by `format.sh`...

Comment (krfricke) [reply src/ray/raylet/raylet.cc:82]: Remove from PR?

Comment (richardliaw) [reply src/ray/raylet/raylet.cc:82]: sure, why not

**Review by richardliaw — APPROVED** [review]

**Pull request merged**